### DISCUSSION TITLE
Public environment variables dont works in kubernetes

### DISCUSSION THREAD

**Discussion by vbpns** [vercel/next.js#25474] [Help] [answered]
Hello guys!

I have some public environment variables (NEXT_PUBLIC_xxxx) in my next.js application.
In my machine works like a charm but when hosted on Kubernetes dont works

Locally: I have a ".env.local" file with the variables (private and public)
Kubernetes: I have a config map and secret map for the environment variables. When i write ENV inside the container all environment variables appears perfect.

But when I try to do a request in the application using a public environment they come off as undefined

I tried a lot of solutions (like generate a ".env" file in deployment/ using publicRuntimeConfig in next.config.js for all public environments) but none of this works

Does anyone know why this is happening?

**Comment by dstarner** [comment]
They get injected at `build` time, so if you are building an image once to be deployed in multiple environments, then it will only see what was available at build time...we are also trying to debug a similar, multi-environment issue.

**Comment by leerob** [comment, marked as answer]
This should be solved with the App Router. We can safely read environment variables on the server during dynamic rendering. This allows you to use a singular Docker image that can be promoted through multiple environments with different values. So you could use whatever environment you want. Build once, run anywhere.

```jsx
import { unstable_noStore as noStore } from 'next/cache';
 
export default function Component() {
  noStore(); // Opt-into dynamic rendering
  // This value will be evaluated at runtime
  const value = process.env.MY_VALUE
  ...
}
```

You can also opt into dynamic rendering in other ways, like using `cookies` or `headers`, but this is an example.

**Reply by ealdwairi** [reply]
Can you plz provide full example?
GitHub ?
